examples/extensions/r2/hello_r2.py

```diff
@@ -33,7 +33,8 @@ def my_sandbox(path, rootfs):
     # get function address and hook it
     ql.hook_address(func, r2.functions['main'].offset)
     # enable trace powered by r2 symsmap
-    r2.enable_trace()
+    # r2.enable_trace()
+    r2.enable_disasm()  # monky-patched disasm powered by r2
```

Comment:
Why not follow eli's advice here?

> My suggestion to monkeypatch the disassemble method meant to replace the default disasembly hook that is set when the user sets the verbosity level to QL_VERBOSE.DISASM. That could have take place when the new r2 module imported, and provide a transparent way to go with the new feature.

Reply:
> That could have take place when the new r2 module imported, and provide a transparent way to go with the new feature. Anyway, the current method is fine too.

I cannot figure a smooth way to call r2 disassembler by setting the verbosity level to `QL_VERBOSE.DISASM` when r2 is not imported by qiling's core.

Reply:
> > That could have take place when the new r2 module imported, and provide a transparent way to go with the new feature. Anyway, the current method is fine too.
> 
> I cannot figure a smooth way to call r2 disassembler by setting the verbosity level to `QL_VERBOSE.DISASM` when r2 is not imported by qiling's core.

I think @elicn means something like:

```python
if verbose == QL_VERBOSE.DISASM:
    # ....
    try:
        from qiling.extension.code import r2
    except ImportError:
        # Monkey patch
```

Reply:
Done. Now when `setup_output` is called (mostly during the initialization of qiling), a instance of R2 will be created, though r2 is not part of qiling's core (`ql.r2` is not possible now)

Reply:
Ah, that is not exactly what I meant..
I meant to replace the `disassembler` method from _outside_, to keep Qiling "clean" from any radare2 references. You can replace the disassembler instance method with something like:
```python
ql.arch.utils.disassembler = another_method.__get__(ql.arch.utils)
```
That can be called within r2 module, and executed when it is loaded. You should not worry about whether the original `disasembler` method was already used for a hook. Just set the verbosity again to itself:
```python
ql.verbose = ql.verbose
```
so `setup_output` will be called and re-install the hook with the new disasembly method.

Reply:
Then how could user enable r2 after r2 installed?

Reply:
I don't understand the question.. can you be more specific?

Reply:
If there is no import in core, how could monkey patch work? Or users have to import…? That doesn’t make too much sense as no one will really write an extra line…

Reply:
Why not..? If someone wants to enable radare2-based disasembly, they should explicitly import the extension - just like any other extension. How is that different from the `trace` extension?

Reply:
> Why not..? If someone wants to enable radare2-based disasembly, they should explicitly import the extension - just like any other extension. How is that different from the `trace` extension?

It's different. `trace` is more of a tool but `r2` will provide many APIs in parallel with some existing Qiling API. Users would always prefer an `out-of-box` solution and we really need some further feedback on r2 integration. I really don't think a few import lines in our files could make so much difference...

Reply:
Are we talking about the disasembly functionality, or is this a wider discussion on radare2-based analysis..?
If the former, than this is just like any other tool - like you said. Anyway, this is only my opinion..

Just thinking out loud here: would it make more sense to create a dedicated branch for radare2-assisted Qiling, to let people give it a spin?

Reply:
> Are we talking about the disassembly functionality, or is this a wider discussion on radare2-based analysis..?
If the former, than this is just like any other tool - like you said. Anyway, this is only my opinion..

> Just thinking out loud here: would it make more sense to create a dedicated branch for radare2-assisted Qiling, to let people give it a spin?

I refer to the wider discussion and I understand your concern. However, I think `pip install qiling[RE]` already provides a "dedicated branch" to a certain extent (thinking of how many users install EVM extensions till today...).

Reply:
@elicn Do you mean a dedicated branch in GitHub where I continuously add new features? Now I can only limited new work until the previous PR get merged since they are not totally independent.

Reply:
I suggested to create a separated dev branch for radare2-assisted Qiling, where radare2 can be integrated without worrying about breaking anything or messing up the code. This is just a continuation of a previous discussion, in which we talked about whether to leave radare2 stuff out of core until Qiling gets re-architected to harness radare2 analysis.

All that being said, I approved this PR a few days ago and as far as I am concerned - it can be merged.

Reply:
Sure, I have made two more new commits including refactor and test to guarantee the API is stable, then this PR can be merged.